# Overview

The main aim of this project is to perform some data cleaning in the [JBeans Raw Dataset](https://drive.google.com/drive/folders/1giaGOhJYWIXfZzy-zxmbrUIOB9Y_ROdu) and get it ready for cleaning 

The main aim of this project is to practice some of the skills I have learnt in data cleaning and data analysis. 


#### Requirements

The directory of this notebook contains a CSV file which contains all the raw data in the Jeans Raw Dataset. If the CSV file is missing you can download it [here](https://drive.google.com/drive/folders/1giaGOhJYWIXfZzy-zxmbrUIOB9Y_ROdu).

The following assumptions have been made:
- The CSV file is in the current directory
- The CSV file is named raw_survey_data.csv




In [2]:
# if raw_survey_data.csv is in the current directory it would be outputed when you run 
# this cell 

!ls | grep raw_survey_data.csv$

raw_survey_data.csv


In [3]:
# importing packages with their natural alias

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

In [4]:

# Here we store all the constants we need in this project
CSV_FILENAME = 'raw_survey_data.csv'


In [120]:
# Loading the CSV to a dataframe
raw_data = pd.read_csv(CSV_FILENAME)


In [121]:
# Performing basic checks on the data 
raw_data.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 19835 entries, 0 to 19834
Columns: 284 entries, Unnamed: 0 to What country do you live in?
dtypes: float64(1), int64(1), object(282)
memory usage: 43.0+ MB


In [7]:
raw_data.columns

Index(['Unnamed: 0',
       'Is Python the main language you use for your current projects?',
       'None:What other language(s) do you use?',
       'Java:What other language(s) do you use?',
       'JavaScript:What other language(s) do you use?',
       'C/C++:What other language(s) do you use?',
       'PHP:What other language(s) do you use?',
       'C#:What other language(s) do you use?',
       'Ruby:What other language(s) do you use?',
       'Bash / Shell:What other language(s) do you use?',
       ...
       'Technical support:Which of the following best describes your job role(s)?',
       'Data analyst:Which of the following best describes your job role(s)?',
       'Business analyst:Which of the following best describes your job role(s)?',
       'Team lead:Which of the following best describes your job role(s)?',
       'Product manager:Which of the following best describes your job role(s)?',
       'CIO / CEO / CTO:Which of the following best describes your job role(s)?

In [8]:

raw_data.shape

(19835, 284)

In [9]:
raw_data.head()

Unnamed: 0 Is Python the main language you use for your current projects?  \
0           1     No, I don’t use Python for my current projects               
1           2     No, I don’t use Python for my current projects               
2           3     No, I don’t use Python for my current projects               
3           4     No, I don’t use Python for my current projects               
4           5     No, I don’t use Python for my current projects               

  None:What other language(s) do you use?  \
0                                    None   
1                                     NaN   
2                                     NaN   
3                                     NaN   
4                                     NaN   

  Java:What other language(s) do you use?  \
0                                     NaN   
1                                     NaN   
2                                     NaN   
3                                    Java   
4                                    Java   

  JavaScript:What other language(s) do you use?  \
0                                           NaN   
1                                    JavaScript   
2                                    JavaScript   
3                                    JavaScript   
4                                           NaN   

  C/C++:What other language(s) do you use?  \
0                                      NaN   
1                                      NaN   
2                                      NaN   
3                                      NaN   
4                                      NaN   

  PHP:What other language(s) do you use?  \
0                                    NaN   
1                                    PHP   
2                                    NaN   
3                                    NaN   
4                                    NaN   

  C#:What other language(s) do you use?  \
0                                   NaN   
1                                   NaN   
2                                   NaN   
3                                   NaN   
4                                   NaN   

  Ruby:What other language(s) do you use?  \
0                                     NaN   
1                                     NaN   
2                                     NaN   
3                                     NaN   
4                                     NaN   

  Bash / Shell:What other language(s) do you use?  \
0                                             NaN   
1                                    Bash / Shell   
2                                    Bash / Shell   
3                                    Bash / Shell   
4                                             NaN   

              ...               \
0             ...                
1             ...                
2             ...                
3             ...                
4             ...                

  Technical support:Which of the following best describes your job role(s)?  \
0                                                NaN                          
1                                                NaN                          
2                                                NaN                          
3                                                NaN                          
4                                                NaN                          

  Data analyst:Which of the following best describes your job role(s)?  \
0                                                NaN                     
1                                                NaN                     
2                                                NaN                     
3                                                NaN                     
4                                                NaN                     

  Business analyst:Which of the following best describes your job role(s)?  \
0                                                NaN                         
1

In [10]:
# List of all columns that were asked

for index, columns in enumerate(raw_data.columns):
    print('#{}: {}'.format(index+1,  columns))
    

#1: Unnamed: 0
#2: Is Python the main language you use for your current projects?
#3: None:What other language(s) do you use?
#4: Java:What other language(s) do you use?
#5: JavaScript:What other language(s) do you use?
#6: C/C++:What other language(s) do you use?
#7: PHP:What other language(s) do you use?
#8: C#:What other language(s) do you use?
#9: Ruby:What other language(s) do you use?
#10: Bash / Shell:What other language(s) do you use?
#11: Objective-C:What other language(s) do you use?
#12: Go:What other language(s) do you use?
#13: Visual Basic:What other language(s) do you use?
#14: Scala:What other language(s) do you use?
#15: SQL:What other language(s) do you use?
#16: Kotlin:What other language(s) do you use?
#17: R:What other language(s) do you use?
#18: Swift:What other language(s) do you use?
#19: Clojure:What other language(s) do you use?
#20: Perl:What other language(s) do you use?
#21: Rust:What other language(s) do you use?
#22: Groovy:What other language(s) do you 

###  Characteristics of the Dataset
We see from the above checks that the:

- dataset has `19,835` rows and `284` columns
- questions asked were placed in columns
- each option in a multi-choice question is stored as a new column in the dataset
- missing values are stored as `NaN`
- once the user says they dont use python as their main language for the first question then the only relevant question for that user is the last 2 namely:

>Could you tell us your age range?

**AND**

>What country do you live in?



### Our goal

Looking at how the data is structured, I think that the best way to structure this data for analysis is to extract the answers of each question and store them in different sheets in an excel file with each sheet having the name of the excel files. 

This, I believe would make it easy to analyze and process the data. 

In summary, we have to do the following:
- each question and extract their answers
- clean data extracted and get it ready for analysis
- export cleaned data into an excel sheet whose sheet name is the question
- save the data to a file

The final output of this notebook therefore is the excel_file. Let us start by creating 2 subsets of the origin dataframe that contains:
- Answers to questions when the user uses python
- Answers to questions when the user does not use python




### Observations subsets

In [59]:
# observations for python files that do not have
NOT_PYTHON_STR = 'No, I don’t use Python for my current projects'

In [48]:
# extracting the observations into two dataframes
NON_PYTHON_OBSERVATIONS_DF = raw_data[raw_data.iloc[:, 1] == NOT_PYTHON_STR]
PYTHON_OBSERVATIONS_DF = raw_data[raw_data.iloc[:, 1] != NOT_PYTHON_STR]


In [62]:
print(NON_PYTHON_OBSERVATIONS_DF.iloc[:, 1].value_counts())
NON_PYTHON_OBSERVATIONS_DF.head()

No, I don’t use Python for my current projects    1404
Name: Is Python the main language you use for your current projects?, dtype: int64


Unnamed: 0 Is Python the main language you use for your current projects?  \
0           1     No, I don’t use Python for my current projects               
1           2     No, I don’t use Python for my current projects               
2           3     No, I don’t use Python for my current projects               
3           4     No, I don’t use Python for my current projects               
4           5     No, I don’t use Python for my current projects               

  None:What other language(s) do you use?  \
0                                    None   
1                                     NaN   
2                                     NaN   
3                                     NaN   
4                                     NaN   

  Java:What other language(s) do you use?  \
0                                     NaN   
1                                     NaN   
2                                     NaN   
3                                    Java   
4                                    Java   

  JavaScript:What other language(s) do you use?  \
0                                           NaN   
1                                    JavaScript   
2                                    JavaScript   
3                                    JavaScript   
4                                           NaN   

  C/C++:What other language(s) do you use?  \
0                                      NaN   
1                                      NaN   
2                                      NaN   
3                                      NaN   
4                                      NaN   

  PHP:What other language(s) do you use?  \
0                                    NaN   
1                                    PHP   
2                                    NaN   
3                                    NaN   
4                                    NaN   

  C#:What other language(s) do you use?  \
0                                   NaN   
1                                   NaN   
2                                   NaN   
3                                   NaN   
4                                   NaN   

  Ruby:What other language(s) do you use?  \
0                                     NaN   
1                                     NaN   
2                                     NaN   
3                                     NaN   
4                                     NaN   

  Bash / Shell:What other language(s) do you use?  \
0                                             NaN   
1                                    Bash / Shell   
2                                    Bash / Shell   
3                                    Bash / Shell   
4                                             NaN   

              ...               \
0             ...                
1             ...                
2             ...                
3             ...                
4             ...                

  Technical support:Which of the following best describes your job role(s)?  \
0                                                NaN                          
1                                                NaN                          
2                                                NaN                          
3                                                NaN                          
4                                                NaN                          

  Data analyst:Which of the following best describes your job role(s)?  \
0                                                NaN                     
1                                                NaN                     
2                                                NaN                     
3                                                NaN                     
4                                                NaN                     

  Business analyst:Which of the following best describes your job role(s)?  \
0                                                NaN                         
1

In [63]:
print(PYTHON_OBSERVATIONS_DF.iloc[:, 1].value_counts())
PYTHON_OBSERVATIONS_DF.head()

Yes                                         15404
No, I use Python as a secondary language     3027
Name: Is Python the main language you use for your current projects?, dtype: int64


Unnamed: 0  \
1404        1405   
1405        1406   
1406        1407   
1407        1408   
1408        1409   

     Is Python the main language you use for your current projects?  \
1404                                                Yes               
1405                                                Yes               
1406                                                Yes               
1407                                                Yes               
1408                                                Yes               

     None:What other language(s) do you use?  \
1404                                     NaN   
1405                                     NaN   
1406                                     NaN   
1407                                     NaN   
1408                                     NaN   

     Java:What other language(s) do you use?  \
1404                                    Java   
1405                                     NaN   
1406                                     NaN   
1407                                     NaN   
1408                                     NaN   

     JavaScript:What other language(s) do you use?  \
1404                                    JavaScript   
1405                                    JavaScript   
1406                                    JavaScript   
1407                                           NaN   
1408                                    JavaScript   

     C/C++:What other language(s) do you use?  \
1404                                      NaN   
1405                                      NaN   
1406                                      NaN   
1407                                      NaN   
1408                                      NaN   

     PHP:What other language(s) do you use?  \
1404                                    NaN   
1405                                    NaN   
1406                                    NaN   
1407                                    NaN   
1408                                    NaN   

     C#:What other language(s) do you use?  \
1404                                   NaN   
1405                                   NaN   
1406                                   NaN   
1407                                   NaN   
1408                                   NaN   

     Ruby:What other language(s) do you use?  \
1404                                     NaN   
1405                                     NaN   
1406                                     NaN   
1407                                     NaN   
1408                                     NaN   

     Bash / Shell:What other language(s) do you use?  \
1404                                             NaN   
1405                                             NaN   
1406                                    Bash / Shell   
1407                                    Bash / Shell   
1408                                             NaN   

                 ...               \
1404             ...                
1405             ...                
1406             ...                
1407             ...                
1408             ...                

     Technical support:Which of the following best describes your job role(s)?  \
1404                                                NaN                          
1405                                                NaN                          
1406                                                NaN                          
1407                                                NaN                          
1408                                                NaN                          

     Data analyst:Which of the following best describes your job role(s)?  \
1404                                                NaN                     
1405                                                NaN                     
1406                                                NaN                     
1407                                              

From the above we have successfully put all observations of python users into the `PYTHON_OBSERVATIONS_DF` constant and all non users into `NON_PYTHON_OBSERVATIONS_DF`


Now we are ready to start processing our questions


## Question 1

In this multi choice question, the users were asked:
> What other language(s) do you use?

The options of this question spans across mulitiple columns number 2 to 25. 


In [74]:
question_one = PYTHON_OBSERVATIONS_DF.iloc[:, 2:26]

In [75]:
question_one

None:What other language(s) do you use?  \
1404                                      NaN   
1405                                      NaN   
1406                                      NaN   
1407                                      NaN   
1408                                      NaN   
1409                                      NaN   
1410                                      NaN   
1411                                      NaN   
1412                                      NaN   
1413                                      NaN   
1414                                      NaN   
1415                                      NaN   
1416                                      NaN   
1417                                      NaN   
1418                                      NaN   
1419                                      NaN   
1420                                      NaN   
1421                                      NaN   
1422                                      NaN   
1423                                      NaN   
1424                                      NaN   
1425                                      NaN   
1426                                      NaN   
1427                                      NaN   
1428                                      NaN   
1429                                      NaN   
1430                                     None   
1431                                      NaN   
1432                                      NaN   
1433                                      NaN   
...                                       ...   
19805                                     NaN   
19806                                     NaN   
19807                                     NaN   
19808                                    None   
19809                                     NaN   
19810                                     NaN   
19811                                     NaN   
19812                                     NaN   
19813                                     NaN   
19814                                     NaN   
19815                                     NaN   
19816                                     NaN   
19817                                     NaN   
19818                                     NaN   
19819                                     NaN   
19820                                    None   
19821                                     NaN   
19822                                     NaN   
19823                                     NaN   
19824                                     NaN   
19825                                     NaN   
19826                                     NaN   
19827                                     NaN   
19828                                     NaN   
19829                                     NaN   
19830                                     NaN   
19831                                     NaN   
19832                                    None   
19833                                     NaN   
19834                                     NaN   

      Java:What other language(s) do you use?  \
1404                                     Java   
1405                                      NaN   
1406                                      NaN   
1407                                      NaN   
1408                                      NaN   
1409                                      NaN   
1410                                     Java   
1411                                     Java   
1412                                      NaN   
1413                                      NaN   
1414                                      NaN   
1415                                      NaN   
1416                                      NaN   
1417                                      NaN   
1418                                      NaN   
1419                                     Java   
1420                                      NaN   
1421                                      NaN   
1422                                

In [107]:
melted_observations = pd.melt(question_one, var_name='option', value_name='reply')

# melted_observations = melted_observations[ melted_observations.reply.dropna()]
melted_observations  = melted_observations.dropna().reset_index()
melted_observations = melted_observations.iloc[:, 1:5]
print(melted_observations.option.value_counts())
print('-----------')
print(melted_observations.reply.value_counts())



JavaScript:What other language(s) do you use?           9233
HTML/CSS:What other language(s) do you use?             8633
Bash / Shell:What other language(s) do you use?         8245
SQL:What other language(s) do you use?                  7451
C/C++:What other language(s) do you use?                5757
Java:What other language(s) do you use?                 4161
PHP:What other language(s) do you use?                  2385
C#:What other language(s) do you use?                   2077
Other - Write In::What other language(s) do you use?    1900
Go:What other language(s) do you use?                   1734
R:What other language(s) do you use?                    1589
TypeScript:What other language(s) do you use?           1385
None:What other language(s) do you use?                 1015
Ruby:What other language(s) do you use?                  800
Visual Basic:What other language(s) do you use?          757
Rust:What other language(s) do you use?                  741
Perl:What other language

In [128]:

melted_observations.reply = melted_observations.reply.astype('category')
melted_observations.option = melted_observations.option.astype('category')
melted_observations.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 60739 entries, 0 to 60738
Data columns (total 2 columns):
option    60739 non-null category
reply     60739 non-null category
dtypes: category(2)
memory usage: 120.3 KB


In [112]:
question_one = melted_observations
question_one.head(20)


option reply
0   None:What other language(s) do you use?  None
1   None:What other language(s) do you use?  None
2   None:What other language(s) do you use?  None
3   None:What other language(s) do you use?  None
4   None:What other language(s) do you use?  None
5   None:What other language(s) do you use?  None
6   None:What other language(s) do you use?  None
7   None:What other language(s) do you use?  None
8   None:What other language(s) do you use?  None
9   None:What other language(s) do you use?  None
10  None:What other language(s) do you use?  None
11  None:What other language(s) do you use?  None
12  None:What other language(s) do you use?  None
13  None:What other language(s) do you use?  None
14  None:What other language(s) do you use?  None
15  None:What other language(s) do you use?  None
16  None:What other language(s) do you use?  None
17  None:What other language(s) do you use?  None
18  None:What other language(s) do you use?  None
19  None:What other language(s) do you use?  None

We have successfully prepared the question one data for analysis and stored it in the `question _one` variable. Next is to process question two


Next let us process the next question 
